sys/net/gnrc/netif/gnrc_netif_device_type.c

```diff
@@ -54,6 +54,81 @@ netopt_t gnrc_netif_get_l2addr_opt(const gnrc_netif_t *netif)
     return res;
 }
 
+#if defined(MODULE_CC110X) || defined(MODULE_NRFMIN)
+static void _create_eui64_from_short(const uint8_t *addr, size_t addr_len,
+                                     eui64_t *eui64)
+{
+    const unsigned offset = sizeof(eui64_t) - addr_len;
+
+    memset(eui64->uint8, 0, sizeof(eui64->uint8));
+    eui64->uint8[3] = 0xff;
+    eui64->uint8[4] = 0xfe;
+    memcpy(&eui64->uint8[offset], addr, addr_len);
+}
+#endif /* defined(MODULE_CC110X) || defined(MODULE_NRFMIN) */
+
+int gnrc_netif_eui64_from_addr(const gnrc_netif_t *netif,
+                               const uint8_t *addr, size_t addr_len,
+                               eui64_t *eui64)
+{
+#if GNRC_NETIF_L2ADDR_MAXLEN > 0
+    if (netif->flags & GNRC_NETIF_FLAGS_HAS_L2ADDR) {
+        switch (netif->device_type) {
+#if defined(MODULE_NETDEV_ETH) || defined(MODULE_ESP_NOW) || \
+    defined(MODULE_NORDIC_SOFTDEVICE_BLE)
+            case NETDEV_TYPE_ETHERNET:
+            case NETDEV_TYPE_ESP_NOW:
+            case NETDEV_TYPE_BLE:
+                if (addr_len == sizeof(eui48_t)) {
+                    eui48_to_eui64(eui64, (const eui48_t *)addr);
```

Comment:
This changes the alignment requirements of `addr`: `const uint8_t *addr` is aligned to one byte, `const eui48_t *addr` is aligned to two bytes.

Reply:
This is already as such in master.

Reply:
(how do I fix it?)

Reply:
(it doesn't lead to problems on the `samr21-xpro` which is quite nitpicky about alignment usually)

Reply:
> This is already as such in master.

See

https://github.com/RIOT-OS/RIOT/blob/a78f357b267d9401ec1d32d9c5561cc55cd0278a/sys/net/gnrc/netif/gnrc_netif_device_type.c#L151

Reply:
> (it doesn't lead to problems on the samr21-xpro which is quite nitpicky about alignment usually)

A random chunk of memory with an alignment requirement has a chance of 50% to also be aligned to two bytes, and a chance of 25% to also alignment to four bytes. This makes those bugs so nasty. Also, some CPUs can handle unaligned memory accesses correctly (maybe at the cost of a performance penalty), while others will simply the `n` least significant bits of the access (with `n == 1` for an alignment of two bytes, `n == 2` for four bytes etc.).

Reply:
The only options to fix I see are:

1. Change the signature of `gnrc_netif_eui64_from_addr` to require `addr` to be aligned to two bytes
2. Change the signature of `eui48_to_eui64` to handle entries aligned to one byte. (The code does so already anyway.)
3. memcpy into a temporary buffer which has adequate alignment

I would favor 2.

(Btw: This bug will never trigger currently, as `eui48_to_eui64` does not make use of the memory alignment of 2 bytes. But a change there could make the bug pop up.)

Reply:
> 2. Change the signature of `eui48_to_eui64` to handle entries aligned to one byte. (The code does so already anyway.)

I'm in favor of that fix as well, however I don't understand where the alignment issue is coming from (sorry, I really have a knowledge gap there). Can I change the definition of `eui48_t` to make it go away?

Reply:
Does #10820 fix this?

Reply:
Basic data types have the alignment requirements of their size. Arrays have the same alignment as their elements. A union and structs will have the same alignment requirement as their member with the highest requirement.

``` C
typedef union {
     uint8_t uint8[6];            // <-- Alignment requirement of 1 bytes
     network_uint16_t uint16[3];  // <-- Alignment requirement of 2 bytes
 } eui48_t;
```

So `eui48_t` will have `MAX(2, 1) == 2` as alignment requirement. Dropping the `uint16` member would reduce the alignment requirement. But I though more about changing `eui48_to_eui64` to accept `uint8_t *` as second argument.

Good to know I'm not the only one still learning while working at RIOT :-)

Reply:
> Does #10820 fix this?

Yep :-)